### PR TITLE
WIP: feature/summarize

### PR DESCRIPTION
**Pull request by Spellchaser**
Porting over summarize

### CODE REVIEW

File: expr/func_summarize.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+	return []Arg{
+		ArgSeriesLists{val: &s.in},
+		ArgString{key: "interval", val: &s.intervalString, validator: []Validator{IsIntervalString}},
+		ArgString{key: "func", opt: true, val: &s.fn, validator: []Validator{IsConsolFunc}},
+		ArgBool{key: "alignToFrom", opt: true, val: &s.alignToFrom},
+	}, []Arg{ArgSeriesList{}}
+}
+
+func (s *FuncSummarize) Context(context Context) Context {
+	context.consol = 0
```
Comment (shanson7) [inline expr/func_summarize.go]: Why reset the consolidation function?

File: expr/func_summarize.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+		out, alignedEnd := summarizeValues(serie, s.fn, interval, newStart, newEnd)
+
+		if s.alignToFrom {
+			newEnd = alignedEnd
+		}
+
+		output := models.Series{
+			Target:     newName(serie.Target),
+			QueryPatt:  newName(serie.QueryPatt),
```
Comment (shanson7) [inline expr/func_summarize.go]: It seems like most functions have `QueryPatt` match the `Target`, but `perSecond` and `scale` do not. Might be worth asking the Metrictank folks what the proper answer is.

File: expr/func_summarize.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+			if s.Datapoints[i].Ts <= ts {
+				s = i
+			}
+			if s.Datapoints[i].Ts >= ts && !math.IsNaN(s.Datapoints[i].Val) {
+				nonNull += 1
+			}
+		}
+
+		// xFilesFactor processing would be implemented here
+		_ = nonNull
```
Comment (shanson7) [inline expr/func_summarize.go]: It doesn't seem like anything in MT uses xFilesFactor. If it did, that would likely be implemented in `aggFunc` so the above should probably be removed for now.

File: expr/func_summarize.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+	// intervalPoints := interval / s.Interval // for xFilesFactor
+
+	numPoints = util.Min(uint32(len(s.Datapoints)), (start-end)/interval)
+
+	ts := start
+	for i := 0; ts < end; ts += interval {
+		s, nonNull := i, 0
+		for ; i < numPoints && s.Datapoints[i].Ts < ts+interval; i++ {
+			if s.Datapoints[i].Ts <= ts {
+				s = i
```
Comment (shanson7) [inline expr/func_summarize.go]: In what scenario does this happen?

File: expr/func_summarize.go (hunk 1 above)
Comment (Spellchaser) [reply expr/func_summarize.go]: https://github.com/bloomberg/metrictank/blob/master/expr/NOTES#L74 says it should be set to default and it's reset in perSecond

File: expr/func_summarize.go (hunk 2 above)
Comment (Spellchaser) [reply expr/func_summarize.go]: I'll stick a comment and ask in the PR

File: expr/func_summarize.go (hunk 4 above)
Comment (Spellchaser) [reply expr/func_summarize.go]: I'm not sure. If the series.start in the Python code is different from QueryFrom, that might be it?

File: expr/func_summarize.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+		alignToFromTarget = ", true"
+	}
+	newName := func(oldName string) string {
+		return fmt.Sprintf("summarize(%s, \"%s\", \"%s\"%s)", oldName, s.intervalString, s.fn, alignToFromTarget)
+	}
+
+	var outputs []models.Series
+	for _, serie := range series {
+		var newStart, newEnd int = serie.QueryFrom, serie.QueryTo
+		if s.alignToFrom {
```
Comment (shanson7) [inline expr/func_summarize.go]: This seems negated from what the docs say.

Comment (Spellchaser) [reply expr/func_summarize.go]: Agreed, my oops

**Comment by Spellchaser** [comment]
Ready for review and testing. I'll refactor the tests so they're smaller later

File: expr/func_summarize.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+		return fmt.Sprintf("summarize(%s, \"%s\", \"%s\"%s)", oldName, s.intervalString, s.fn, alignToFromTarget)
+	}
+
+	var outputs []models.Series
+	for _, serie := range series {
+		var newStart, newEnd uint32 = serie.QueryFrom, serie.QueryTo
+		if len(serie.Datapoints) > 0 {
+			newStart = serie.Datapoints[0].Ts
+			newEnd = serie.Datapoints[len(serie.Datapoints)-1].Ts + serie.Interval
+		}
```
Comment (Spellchaser) [inline expr/func_summarize.go]: Otherwise if no data points are returned...

File: expr/func_summarize.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+			newEnd = newEnd - (newEnd % interval) + interval
+		}
+
+		out, alignedEnd := summarizeValues(serie, s.fn, interval, newStart, newEnd)
+
+		if s.alignToFrom {
+			newEnd = alignedEnd
+		}
+
+		output := models.Series{
```
Comment (Spellchaser) [inline expr/func_summarize.go]: Graphite seems to set QueryFrom, QueryTo = newStart, newEnd here if series.start is equivalent to QueryFrom and series.end " QueryTo

File: expr/func_summarize.go [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+			Target:     newName(serie.Target),
+			QueryPatt:  newName(serie.QueryPatt), // Does this exist?
```
[the remark this answers is not in the document]
Comment (Spellchaser) [reply expr/func_summarize.go]: To ask the grafana people
https://github.com/bloomberg/metrictank/pull/15#discussion_r163061745

[the remark this answers is not in the document]
Comment (shanson7) [reply expr/func_summarize.go]: Ah, that's a very misleading comment. More like "Should this match target?"

File: expr/func_summarize.go [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+	if s.alignToFrom {
+		alignToFromTarget = ", true"
+	}
+	newName := func(oldName string) string {
+		return fmt.Sprintf("summarize(%s, \"%s\", \"%s\"%s)", oldName, s.intervalString, s.fn, alignToFromTarget)
+	}
+
+	var outputs []models.Series
+	for _, serie := range series {
+		var newStart, newEnd uint32 = serie.QueryFrom, serie.QueryTo
```
[the remark this answers is not in the document]
Comment (Spellchaser) [reply expr/func_summarize.go]: From the grafana query inspector, I wasn't sure how to see the from and to or what functions would alter it

File: expr/func_summarize.go [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+		outputs = append(outputs, output)
+		cache[Req{}] = append(cache[Req{}], output)
+	}
+	return outputs, nil
+}
+
+func summarizeValues(serie models.Series, fn string, interval, start, end uint32) ([]schema.Point, uint32) {
+	out := pointSlicePool.Get().([]schema.Point)
+
+	aggFunc := consolidation.GetAggFunc(consolidation.FromConsolidateBy(fn))
```
[the remark this answers is not in the document]
Comment (Spellchaser) [reply expr/func_summarize.go]: Ahh I was thinking the agg function could be different for each series before I played with grafana

**Pull request merged**